### PR TITLE
(Feature) #715 protect endpoints

### PR DESCRIPTION
**Pull request by yaroslav-fedyshyn-nordwhale**
# Description

Add verification whether the provided email already exists in private user DB for POST /user/add request.
Add rate limit for send email/phone code requests.

Fixes https://app.zenhub.com/workspaces/gooddollar-5d50833888a83c6880d3e345/issues/gooddollar/gooddapp/715

# How Has This Been Tested?

It should be tested on Dapp (http://localhost:3000).
The instructions for testing is described in Dapp PR.
You can find link to Dapp PR below Checklist.

# Checklist:
- [x] PR title matches follow: (Feature|Bug|Chore) Task Name
- [x] My code follows the style guidelines of this project
- [x] I have followed all the instructions described in the initial task (check Definitions of Done)
- [x] I have performed a self-review of my own code
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] I have added reference to a related issue in the repository
- [x] I have added a detailed description of the changes proposed in the pull request. I am as descriptive as possible, assisting reviewers as much as possible.
- [ ] I have added screenshots related to my pull request ( for frontend tasks)
- [ ] I have pasted a gif showing the feature.
- [x] @mentions of the person or team responsible for reviewing proposed changes

Link to Dapp PR: https://github.com/GoodDollar/GoodDAPP/pull/731

### CODE REVIEW

File: jest.config.js [hunk 1]
```diff
@@ -152,7 +160,15 @@ module.exports = {
...
-    '/src/server/server.config.js'
+    '/src/server/server.config.js',
+    'src/server/blockchain',
+    'src/server/db',
+    'src/server/gun',
+    'src/server/loadtest',
+    'src/server/login',
+    'src/server/mautic',
+    'src/server/send',
+    'src/server/storage',
```
Comment (sirpy) [inline jest.config.js]: what is all these?

File: src/server/verification/__tests__/verificationAPI.js [hunk 2]
```diff
@@ -63,6 +63,15 @@ describe('verificationAPI', () => {
       .expect(200, { ok: 1, onlyInEnv: { current: 'test', onlyIn: ['production', 'staging'] } })
   })
 
+  test('/verify/sendotp should fail with 429 status - too many requests (rate limiter)', async () => {
+    // the 4th request should fail with status 429
```
Comment (sirpy) [inline src/server/verification/__tests__/verificationAPI.js]: make requests here until you get 429

File: src/server/verification/__tests__/verificationAPI.js [hunk 3]
```diff
@@ -93,6 +102,37 @@ describe('verificationAPI', () => {
...
+      identifier: '0x7ac080f6607405705aed79675789701a48c76f55',
+      fullName: 'test_user_sendemail'
+    })
+
+    expect(user).toBeTruthy()
+
+    // the 4th request should fail with status 429
+    // we see only 3 requests executed from this test. But in previous test we executing one more api call to the same endpoint.
+    // so in summary we have 4 requests
+    for (let i = 1; i <= 3; i++) {
```
Comment (sirpy) [inline src/server/verification/__tests__/verificationAPI.js]: loop until you get 429 dont use fixed number

File: src/server/utils/requestRateLimiter.js [hunk 4]
```diff
@@ -0,0 +1,7 @@
+import rateLimit from 'express-rate-limit'
+
+export default (t = 1, n = 3) =>
+  rateLimit({
+    windowMs: +t * 60 * 1000, // minutes
+    max: +n // limit each IP to n requests per windowMs
```
Comment (sirpy) [inline src/server/utils/requestRateLimiter.js]: read n,t from env variable

File: src/server/storage/storageAPI.js [hunk 5]
```diff
@@ -39,8 +39,14 @@ const setup = (app: Router, storage: StorageAPI) => {
       )
         throw new Error('User email or mobile not verified!')
 
-      const { email, mobile, ...bodyUser } = body.user
-      const user: UserRecord = defaults(bodyUser, {
+      const { email } = body.user
+      const isDuplicatedEmail = await storage.isDupUserData({ email })
```
Comment (sirpy) [inline src/server/storage/storageAPI.js]: this is going to fail for every new user.
you need to check if createdDate exists, thats the parameter we set when we first add user

Comment (yaroslav-fedyshyn-nordwhale) [reply src/server/storage/storageAPI.js]: @sirpy I don't think so. It just checking if there is record with provided email in user collection.

File: jest.config.js (hunk 1 above)
Comment (yaroslav-fedyshyn-nordwhale) [reply jest.config.js]: done. removed it. I added this to run specific tests locally. And pushed it by mistake.

File: src/server/verification/__tests__/verificationAPI.js (hunk 2 above)
Comment (yaroslav-fedyshyn-nordwhale) [reply src/server/verification/__tests__/verificationAPI.js]: done. see new commits.

File: src/server/verification/__tests__/verificationAPI.js (hunk 3 above)
Comment (yaroslav-fedyshyn-nordwhale) [reply src/server/verification/__tests__/verificationAPI.js]: done. see new commits.

File: src/server/utils/requestRateLimiter.js (hunk 4 above)
Comment (yaroslav-fedyshyn-nordwhale) [reply src/server/utils/requestRateLimiter.js]: done. see new commits.

File: src/server/storage/__tests__/storageAPI.js [hunk 6]
```diff
@@ -38,7 +38,7 @@ describe('storageAPI', () => {
     await UserDBPrivate.updateUser({ identifier: '0x7ac080f6607405705aed79675789701a48c76f55', mauticId: mauticId })
     const user: UserRecord = {
       identifier: '0x7ac080f6607405705aed79675789701a48c76f55',
-      email: 'useraddtest@gooddollar.org' // required for mautic create contact
+      email: 'useraddtest1@gooddollar.org' // required for mautic create contact
```
Comment (sirpy) [inline src/server/storage/__tests__/storageAPI.js]: what is this fix?

File: src/server/storage/storageAPI.js (hunk 5 above)
Comment (sirpy) [reply src/server/storage/storageAPI.js]: user email field was set when user verified his email, so this WILL fail once he calls /user/add.
Just replace check with the createdDate
Stas added a fix that email/mobile will be saved only in otp record and set in /user/add so after his fix this might not fail. But still the correct way is just to check the userrecord date created, if it exists then user can't call /user/add again

File: src/server/storage/__tests__/storageAPI.js (hunk 6 above)
Comment (yaroslavfedyshyn) [reply src/server/storage/__tests__/storageAPI.js]: @sirpy I will revert this. as we change the duplication account check flow.

File: src/server/storage/storageAPI.js (hunk 5 above)
Comment (yaroslavfedyshyn) [reply src/server/storage/storageAPI.js]: done

File: src/server/storage/__tests__/storageAPI.js (hunk 6 above)
Comment (yaroslav-fedyshyn-nordwhale) [reply src/server/storage/__tests__/storageAPI.js]: done

File: src/server/storage/storageAPI.js (hunk 5 above)
Comment (yaroslav-fedyshyn-nordwhale) [reply src/server/storage/storageAPI.js]: done

**Comment by yaroslav-fedyshyn-nordwhale** [comment]
> The build fails because the change causes unit test to return different value
> 
> 1. add unit test to check the new functionality (ie duplicate account)
> 2. modify the failing unit test to check correctly what it tested before

done
